rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,292 @@
+operator:
+
+- `vast pull` → `from vast://...`
+- `vast push` → `to vast://...`
+
+An `export` would then become:
+
+```bash
+vast pull 'where EXPRESSION | print json'
+# Same with exec:
+vast exec 'from vast://.... | where EXPRESSION | print json'
+# Same with export:
+vast export json EXPRESSION
+```
+
+An `import` degenerates to:
+
+```bash
+vast push 'parse zeek | where EXPRESSION'
+# Same with exec
+vast exec 'parse zeek | where EXPRESSION | to vast://...'
+# Same with import
+vast import zeek EXPRESSION
+```
+
+### Mutating existing data
+
+Security telemetry data is often messy. After the data collection aspects have
+been sorted out and analysts interact with the data, it becomes clear that some
+part of the data needs to be cleaned. This matches ELT mindset of dumping
+everything first so that the data is there, and then tweak as you go.
+
+Example operations include, migrating data to a different schema, enriching
+telemetry (e.g., with GeoIP data), and fixing invalid entries.
+
+Mutation of data is non-trivial in that it requires exclusive access to the
+data. Otherwise data races may occur. Consequently, we need transactional
+interface (ACID) to support this operation.
+
+As long as VAST owns the underlying data (i.e., only VAST is allowed to make
+changes), VAST can already mutate of data at rest. [Spatial
+compaction][mutate-at-rest] uses a pre-defined disk quota as trigger to apply
+pipelines to a subset of to-be-transformed data. After VAST applies the
+pipeline, VAST optionally removes the original data in an atomic fashion.
+
+[mutate-at-rest]: https://vast.io/docs/use-vast/transform#modify-data-at-rest
+
+Modeled after the `compaction` plugin, we may consider exposing a mutable
+pipeline interface through a dedicate command. This command interacts with a
+node, and could be seen as a combination of `pull` and `push` that executes
+remotely. For example:
+
+```bash
+vast mutate --types=A,B,C 'where 10.0.0.0/8 | put orig.h resp.h'
+```
+
+This would apply the pipeline to filter and project in schema types `A`, `B`,
+and `C`.
+
+## Implementation
+
+We could implement the entire pipeline with Arrow IPC streams in the "interior"
+operators, with only `from` and `to` taking raw input as byte stream. But this
+may be too restrictive. Let's assume that input and output are dynamically
+typed. An operator must then define its input and output types, e.g., `from<In,
+Out>` takes as input `In` and produces instances of `Out`.
+
+### Matcher Example
+
+This can come in handy for working with the matcher plugin. Consider the use
+case of constructing a matcher from data. Consider this pipeline:
+
+```bash
+vast exec 'read csv |
+           matcher build --extract=net.src.ip' > ips.flatbuffer
+```
+
+Here, we have:
+
+- `read<Bytes, Arrow>`: reads `Bytes` and produces `Arrow` record batches
+- `matcher build<Arrow, Flatbuffer>`: consumes `Arrow` and create a matcher
```

Comment:
This kind of directly ties into my request to want to have a complete list of all possible alternatives that can be input or output types. I think this will heavily influence how we design our operators, so we should come up with that list early.

Reply:
ah yes, sorry I didn't get that at first reading because your phrasing was very C++ oriented 😄

Reply:
@rdettai I rewrote the matcher paragraph to address your top-level concern. Closing this thread, as we can continue in the new content.